### DISCUSSION TITLE
Teach "modern" alternatives to some git commands

### DISCUSSION THREAD

**Discussion by alexdewar** [ImperialCollegeLondon/rse_introductory_git_course#61] [Ideas]
Over the years, the git devs have added some new commands/aliases to make the CLI for git a bit easier to understand. The reason I put scare quotes around "modern" is because the old versions aren't deprecated or anything -- but in any case I'm wondering whether we should use these newer forms in the courses to make it easier for students to wrap their heads around git concepts.

For example, there is now `git stage` as a built-in alias for `git add`, which is somewhat clearer. The concept of "adding" things is a bit ambiguous in the git context (it sounds like you're committing it to the repo) and could be confusing for newbies. Furthermore, you might think that the `git rm` command does the opposite -- i.e. unstages files -- but it deletes them instead. (It's also confusing that you use `git reset` to unstage files, when `git reset` is also used in other contexts. There's now a `git restore` command which *just* unstages files.)

The `git checkout` command is also ambiguous as it can be used to change branches or reset files, so there is now a `git switch` command which *just* changes branches.

I made this a discussion in the intro repo because we probably want to teach this consistently from the start, if this is what we want to do.

Thoughts?

**Comment by AdrianDAlessandro** [comment]
I think using the more modern names makes sense. It's easy to add a line referencing the old one (only really needed for `add` and `checkout`).

One correction: `git restore` by itself only operated on changes in unstaged files, you need to use `git restore --staged` to unstage a file. This is pretty easy to get right though because `git status` will tell you what to do.

**Reply by alexdewar** [reply]
Ah ok. Good to know!

**Comment by dc2917** [comment]
I wasn't aware of `git stage`, but I do find the staging (vs. just committing) aspect to be something newcomers find particularly confusing, so it would probably be clearer if the command at least has the same name as what we say it does. "Use `git stage` to stage" certainly sounds more logical than "Use `git add` to stage". Still, it only takes about two commits to get used to it.

I think that we should maybe avoid saying something like "Alternatively, you can also use `git add` to stage", since this is right at the start when we're teaching the first few commands, and I think this could easily cause confusion/frustration if we're throwing loads of commands at them, some of which we don't expect them to use.

On `checkout` vs `switch`: we have had people ask about checkout vs switch before. I also had someone super confused about checkout meaning to "check out" of a branch like you check out of a hotel, i.e. you're no longer on that branch. (When I first started with version control and was told to check out a branch I thought that meant no one else would be able to work on it, like checking out a book from the library 😅). I'm in favour of teaching `switch` to change branches, but if we do use `checkout` to do things other than changing branches then we should maybe explain a bit more about what `checkout` does for clarity.

**Reply by alexdewar** [reply]
> I think that we should maybe avoid saying something like "Alternatively, you can also use `git add` to stage", since this is right at the start when we're teaching the first few commands, and I think this could easily cause confusion/frustration if we're throwing loads of commands at them, some of which we don't expect them to use.

I agree we don't want to confuse students by telling them to use either. We should tell them to use `git stage`, but just mention that elsewhere (on StackOverflow or whatever) they might see `git add`, which is the same thing.

> On `checkout` vs `switch`: we have had people ask about checkout vs switch before. I also had someone super confused about checkout meaning to "check out" of a branch like you check out of a hotel, i.e. you're no longer on that branch. (When I first started with version control and was told to check out a branch I thought that meant no one else would be able to work on it, like checking out a book from the library 😅). I'm in favour of teaching `switch` to change branches, but if we do use `checkout` to do things other than changing branches then we should maybe explain a bit more about what `checkout` does for clarity.

I've just looked at this now, but it seems you can use `git switch` to check out any old commit if you want to, **BUT** you have to pass the `--detach` flag to do so, which is also clearer! (If you forget it, git will tell you you need to do this.)

**Comment by dc2917** [comment]
So something occurred to me regarding `git add` vs `git stage`.

When you have untracked files or unstaged changes, the output of `git status` says 

```sh
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   <file>
```

```sh
Untracked files:
  (use "git add <file>..." to include in what will be committed)
	<file>
```
and

```sh
no changes added to commit (use "git add" and/or "git commit -a")
```

i.e. `git` always says `git add`, not `git stage`.

When I'm delivering the course I always point this out and encourage them to read the output and take note of the hints that `git` gives you.

Do we still think it's best to go along with `git stage` and just make sure they know it does the same thing as `git add`?

**Reply by alexdewar** [reply]
I had clocked that... it is a little unfortunate. I would still point it out, but make it clear that it's exactly the same as `git stage`. I suppose they should probably know that `git add` is a thing, just in case they come across it in tutorials/StackOverflow/whatever.
